Add debian packaging and add native IPC implementation

pgorgon-hem opened the pull request:
---

### Overview

This pull request introduces native IPC transport implementations and debian (.deb) packages.

### Details

- **Core IPC Updates:**
  - Replaced `parity-tokio-ipc` with platform-specific transport layers for Unix sockets and Windows named pipes.
  - Introduced support for socket activation, security attributes, and updated related APIs and examples.

- **Linux Packaging:**
  - Added Debian packaging scripts to improve service management (systemd user and system services), dependency checks, and changelog generation.
  - Streamlined service handling through updates to systemd configurations and service maintenance scripts for better portability and reliability.

- **Testing Improvements:**
  - Captured detailed child process logs during integration tests for better debugging.
  - Added comprehensive tests for multi-client IPC compatibility.

--- 

### Checklist

- [ ] Tests added/updated for new features.
- [ ] Documentation updated to reflect changes.
- [ ] Changes reviewed for potential regressions.

pjgorgon committed:
feat(linux): add debian package for FSCT host services

Introduced a Debian package for FSCT host services, including the system-wide driver and per-user client integration. Added scripts for installation, removal, and service management, as well as systemd units for driver and user services.

pjgorgon committed:
feat(linux): update FSCT host service packaging and systemd integration

Refined Debian package for FSCT host services. Updated package metadata, removed unused socket handling, and reorganized scripts for simplified installation and removal. Enhanced systemd service configurations with socket activation support and resource management improvements.

pjgorgon committed:
feat(linux): refine Debian packaging and systemd service configuration

Updated the build script to improve package version retrieval and adjusted the build process. Enhanced systemd service configurations with additional dependencies, a refined `ExecStart` command, and improved service dependency management. Simplifies packaging and service reliability.

pjgorgon committed:
feat(linux): rename FSCT host to FSCT driver and update packaging

Renamed the Debian package and services from FSCT host to FSCT driver, aligning with updated functionality. Modified installation scripts, service units, and build processes to reflect the new naming convention. Renamed user service to fsct-mpris.service to better align with functionality it provides.

pjgorgon committed:
feat(linux): add initial native packaging for FSCT driver

Introduced Debian packaging for the FSCT driver, including debhelper integration, systemd service units, and build enhancements. Streamlined installation process with support for system and user services.

pjgorgon committed:
feat(linux): enhance Debian packaging with improved build script and dependency checks

Refined `build_linux_deb.sh` script to add dependency preflight checks, out-of-source build support, and new flags (`--allow-missing-deps`, `--keep-build`). Updated package rules to use `--locked` flag for Cargo and adjusted `debian/control` for compatibility improvements. Simplifies build process and improves maintainability.

pjgorgon committed:
feat(linux): improve Debian packaging with dynamic versioning and streamlined systemd handling

Enhanced `build_linux_deb.sh` to dynamically determine package version from `Cargo.toml` and update the Debian changelog. Streamlined systemd user service handling by replacing manual symlink creation with `dh_installsystemduser`. Simplifies build and installation processes.

pjgorgon committed:
feat(linux): add maintainer scripts for systemd user service management

Introduced `postinst`, `prerm`, and `postrm` maintainer scripts to manage the `fsct-mpris.service` systemd user service. Scripts handle service restart, stop, and daemon-reload actions for logged-in users during package installation, upgrade, and removal, ensuring smoother user experience.

pjgorgon committed:
fix(linux): update postinst script for improved service handling

Replaced conditional `restart` commands with `enable --now` in the postinst script. Ensures the `fsct-mpris.service` is correctly started and enabled during installation or upgrade, enhancing reliability.

pjgorgon committed:
fix(linux): ensure service management scripts are non-interactive and resilient

Updated `postinst`, `prerm`, and `postrm` scripts to enforce non-interactive, best-effort behavior during systemd user service management. Improved handling for active user sessions and suppressed unnecessary outputs, ensuring smoother script execution.

pjgorgon committed:
feat(linux): enhance Debian packaging and streamline service management

Added `dh-exec` dependency and refactored maintainer scripts (`postinst`, `prerm`, `postrm`) for improved clarity and robustness. Updated systemd service units for better alignment, introduced `.install` files, and optimized the build script for dependency handling and packaging consistency.

pjgorgon committed:
feat(core): add socket activation support and UDS client utility

Implemented socket activation support for Unix systems in IPC server, enabling integration with systemd or other activation helpers. Added a simple Unix Domain Socket (UDS) client example for debugging and manual testing. Included new dependencies (`nix`, `uuid`, and `which`) and updated corresponding configurations.

pjgorgon committed:
feat(native): integrate msgpack-rpc for IPC and enhance testing

Added `msgpack-rpc` for enhanced inter-process communication and introduced `parity-tokio-ipc` for client-server interactions. Updated tests to include a simulated client performing RPC operations, ensuring compatibility with new socket activation behavior. Introduced required dependencies and features to support changes.

pjgorgon committed:
feat(native): capture child process logs in integration tests

Added support for capturing stdout and stderr logs of child processes during integration tests. Introduced pipes and reader threads to handle log forwarding. Updated test behavior to include detailed logs for debugging and ensured proper cleanup of resources. Enhanced `nix` dependency to include `fs` feature to support changes.

pjgorgon committed:
feat(native): remove unused UDS client example and improve test readability

Deleted the outdated UDS client example to reduce code clutter, as it is no longer needed for testing or debugging purposes. Renamed the socket activation integration test for better clarity and to accurately reflect its function.

pjgorgon committed:
refactor(native): simplify pipe file descriptors in integration test

Replaced verbose variable definitions for pipe file descriptors with streamlined naming to improve code readability and clarity. Retained original behavior and ensured no functional changes were introduced.

pjgorgon committed:
feat(core): enhance IPC server with socket activation and API updates

Introduced socket activation support for the IPC server, enabling seamless integration with systemd. Updated the API to include `run_ipc_server_with_fd` and revised functions for endpoint management. Refactored tests and examples to align with the new changes.

pjgorgon committed:
refactor(core): improve IPC server endpoint handling

Refactored `serve` function in IPC server to enforce clearer endpoint type branching logic. Added explicit returns for better readability and reliability. Introduced error handling for invalid endpoint definitions and ensured behavior consistency across conditional branches.

pjgorgon committed:
feat(core): replace `parity-tokio-ipc` with platform-native transports

Replaced the `parity-tokio-ipc` dependency with custom-built platform-native transport abstractions for Unix sockets and Windows named pipes. Refactored client and server implementations to use the new transport layers, improving portability and maintainability. Added comprehensive tests to verify compatibility with multiple clients.

pjgorgon committed:
feat(core): extend IPC transport with security and system compatibility

Enhanced IPC transport by introducing platform-specific security attributes. Added permissive DACL for Windows and updated Unix security attributes API for parity. Refactored transport listener handling to improve error management and compatibility across systems, ensuring robust and secure endpoint creation.

pjgorgon committed:
refactor(core): unify IPC server endpoint handling logic

Consolidated and streamlined endpoint handling in the IPC server by introducing a unified listener initialization method. Improved readability and maintainability by reducing redundancy and enforcing consistency in endpoint branching and error handling.

pjgorgon committed:
refactor(native): remove redundant `fsct-driver.service` enablement

Removed unnecessary enablement of `fsct-driver.service` in the Debian post-install script. Streamlined the configuration process by retaining only the essential `fsct-driver.socket` activation.

pjgorgon committed:
fix(native): ensure correct service stop and disable order in Debian scripts

Resolved service stop and disable sequence issues in Debian `prerm` and `postrm` scripts. Updated to correctly handle `fsct-driver.service` alongside `fsct-driver.socket` to avoid potential race conditions or inconsistencies during package removal or upgrade.

pjgorgon committed:
fix(native): update service dependencies and targets in FSCT user service

Aligned `fsct-driver.user.service` dependencies by adding `graphical-session.target` and adding dependency on `dbus.service`. Improved service behavior and ensured correct integration with DBus for MPRIS functionality.

pjgorgon committed:
fix(native): handle terminate signal in service shutdown logic

Updated the service shutdown logic to handle `SIGTERM` alongside `SIGINT` using a `tokio::select!` construct. Ensured graceful shutdown behavior when receiving termination signals on Unix platforms.

pjgorgon committed:
feat(core): add `wait_for_any_to_finish` and enhance service shutdown handling

Introduced `wait_for_any_to_finish` in the core service for monitoring task completions. Improved shutdown logic by handling task errors, logging failures, and ensuring graceful termination. Enhanced signal handling to await service completion before shutdown.

jarekhem commented on core/src/ipc/server.rs:
```diff
@@ -44,39 +44,63 @@ use std::future::Future;
+    pub async fn serve(&mut self) -> anyhow::Result<()> {
+        let endpoint = self.endpoint.take().expect("serve() called after shutdown");
+
+        let listener = match endpoint {
```
Może wywalić to do jakiejś funkcji typu create_listener czy coś?

jarekhem commented on core/src/ipc/server.rs:
```diff
@@ -138,13 +159,17 @@ impl IpcServer {
+#[cfg(unix)]
+use std::os::fd::{AsRawFd, FromRawFd, OwnedFd};
+#[cfg(unix)]
+use std::os::fd::IntoRawFd;
```
Wrzuć to do tej grupy w 163 linijce

pjgorgon committed:
refactor(core): extract listener initialization in IPC server

Moved the listener initialization logic into a separate `init_listener` method, reducing redundancy within the `serve` function. Replaced synchronous file operations with asynchronous equivalents using `tokio::fs` for better performance and consistency.

pjgorgon committed:
refactor(core): reorganize imports and streamline IPC server setup

Reorganized imports in `ipc/server.rs` for better clarity and structure. Updated `init_listener` function to specify the `transport::EndpointListener`, enhancing code readability and maintainability. Removed redundant and unused imports.

pjgorgon committed:
feat(core): refactor IPC server and unify endpoint listener logic

Simplified IPC server and listener logic by removing redundant endpoint wrapping and consolidating methods for both Unix and Windows platforms. Streamlined the initialization by introducing direct `EndpointListener::from_path/from_fd` methods and reducing unnecessary resource cleanup in tests. Improved error handling and consistency in transport listener creation across all IPC implementations.

pjgorgon committed:
feat(native): migrate to FPM-based packaging for fsct-driver

Replaced Debian-specific packaging files and processes with FPM-based configuration for enhanced flexibility and maintainability. Updated the build script to generate Debian packages using FPM and reorganized systemd-related files. Added support for optional license generation and improved build script helper options.

pjgorgon committed:
feat(build): add detection of glibc version and enforce libc6 dependency

Enhanced the build script to automatically detect the system's glibc version and add a corresponding libc6 dependency in the generated Debian package. Introduced fallback mechanisms to ensure robust detection. Updated Cargo configuration to set panic behavior to "abort" for improved optimization.

pjgorgon committed:
refactor(core): remove dependency on parity-tokio-ipc

Dropped the usage of `parity-tokio-ipc` in favor of native `UnixStream` and `windows pipes` for transport in the IPC server. Updated related imports, configuration, and tests. Streamlined code to reduce redundant dependencies and enhance maintainability.
